### PR TITLE
Changes for making netx.ws thread-safe

### PR DESCRIPTION
**Pull request by sanjay-saxena**
Thread-safe implementation of netx.ws that is garbage-free and wait-free.

### CODE REVIEW

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java [hunk 1]
```diff
@@ -108,6 +109,9 @@
     private final WebSocketInputStateMachine inputStateMachine;
     private final WebSocketOutputStateMachine outputStateMachine;
     private final WebSocketExtensionFactory extensionFactory;
+    private final OptimisticReentrantLock readLock;
+    private final OptimisticReentrantLock stateLock;
+    private final OptimisticReentrantLock writeLock;
 
     private String negotiatedProtocol;
     private WsInputStream inputStream;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java]: Do these non-final fields need to be `volatile` to make sure different threads see the initialized values even if they are racing?

Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java]: Since the `happens-before` relationship will kick-in only after the second thread acquires the lock released by the first thread, I think you are right.

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java [hunk 2]
```diff
@@ -288,4 +291,18 @@ else if (reasonLength > 123) {
         }
     }
 
+    private void createMask(byte[] maskBuf) {
+        int m = maskInt.incrementAndGet();
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: Is the mask predictable?  If so, then a malicious application can control the bytes on the wire, whereas the mask is intended to prevent such predictability.

File: src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java [hunk 3]
```diff
@@ -51,6 +52,7 @@
     private final InputStream in;
     private final FrameRW incomingFrame;
     private final FrameRO incomingFrameRO;
+    private final OptimisticReentrantLock stateLock;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java]: Should this just be of type `Lock` ?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsOutputStream.java [hunk 4]
```diff
@@ -40,6 +41,7 @@
     private final FrameRW outgoingControlFrame;
     private final FrameRO outgoingFrameRO;
     private final ByteBuffer heapBufferControlFrameRO;
+    private final OptimisticReentrantLock stateLock;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsOutputStream.java]: Type `Lock`?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsReader.java [hunk 5]
```diff
@@ -57,6 +58,7 @@
     private final InputStream in;
     private final FrameRW incomingFrame;
     private final FrameRO incomingFrameRO;
+    private final OptimisticReentrantLock stateLock;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsReader.java]: Type `Lock` ?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsReader.java [hunk 6]
```diff
@@ -155,97 +157,104 @@ public int read(char[] cbuf, int offset, int length) throws IOException {
...
+                        // networkBuffer needs to be resized.
+                        int additionalBytes = Math.max(BUFFER_CHUNK_SIZE, payloadLength);
+                        byte[] netBuffer = new byte[networkBuffer.length + additionalBytes];
                         int len = networkBufferWriteOffset - networkBufferReadOffset;
-                        System.arraycopy(networkBuffer, networkBufferReadOffset, networkBuffer, 0, len);
+
+                        System.arraycopy(networkBuffer, networkBufferReadOffset, netBuffer, 0, len);
+                        networkBuffer = netBuffer;
+                        heapBuffer = ByteBuffer.wrap(networkBuffer);
+                        heapBufferRO = heapBuffer.asReadOnlyBuffer();
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsReader.java]: Are these allocations in the message path?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsWriter.java [hunk 7]
```diff
@@ -36,8 +37,9 @@
     private final WsURLConnectionImpl connection;
     private final FrameRW outgoingFrame;
     private final FrameRO outgoingFrameRO;
-    private ByteBuffer payload;
+    private final OptimisticReentrantLock stateLock;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsWriter.java]: Type `Lock`?

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            stamp.incrementAndGet();
+            return;
+        }
+
+        while (!lockObtained) {
+            if (reference.compareAndSet(null, Thread.currentThread())) {
+                stamp.set(1);
+                lockObtained = true;
+            }
+        }
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java:50]: Can this be simplified to the following?

``` java
public void lock() {
  Thread currentThread = Thread.currentThread();
  while (reference.compareAndSet(null, currentThread) == false);
  stamp.set(1);
}}
```

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+package org.kaazing.netx.ws.internal.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+// ### TODO: 1. happens-before relationship should be established between the thread that releases the lock and the next thread
+//              that acquires the lock.
+//           2. Evaluate using Martin's sun.misc.Unsafe based AtomicSequence implementation.
+public class OptimisticReentrantLock {
+    private final AtomicReference<Thread> reference;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: Call this one `owner` instead of `reference`?

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+            if (currentStamp == 1) {
+                // Order in which atomics are reset is important. First, reset the stamp and then the reference.
+                stamp.set(0);
+                reference.set(null);
+            }
+            else if (currentStamp > 1) {
+                stamp.decrementAndGet();
+            }
+        }
+    }
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java:94]: Can this be simplified to the following?

``` java
public void unlock() {
  Thread currentThread = Thread.currentThread();
  if (currentThread == reference.get() && stamp.decrementAndGet() == 0) {
    // order in which atomics are updated is important
    // stamp MUST be zero before nulling the owner
    // defeat unlock / lock race by setting owner to null only if not already updated
    reference.compareAndSet(currentThread, null);
  }
}
```

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java (hunk 8 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java:50]: It has to be like this:

``` java
public void lock() {
  Thread currentThread = Thread.currentThread();
   if (Thread.currentThread() == getOwner()) {
       stamp.incrementAndGet();
       return;
   }
  while (reference.compareAndSet(null, currentThread) == false);
  stamp.set(1);
}}
```

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java (hunk 2 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: Good point. I will change it to

``` java
int m = 1 + random.nextInt(Integer.MAX_VALUE);
```

We are moving away from SecureRandom.nextBytes() as it is creating garbage.

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ */
+package org.kaazing.netx.ws.internal.util;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+// ### TODO: 1. happens-before relationship should be established between the thread that releases the lock and the next thread
+//              that acquires the lock.
+//           2. Evaluate using Martin's sun.misc.Unsafe based AtomicSequence implementation.
+public class OptimisticReentrantLock {
```
Comment (jitsni) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: I haven't checked the usages, but it doesn't guarantee fairness (a thread that comes first doesn't get first) and some threads will be keep on spinning until lock is available.

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java (hunk 8 above)
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java:50]: Sorry, I originally didn't have the method signature around it, then omitted the leading check accidentally.

You probably meant to reuse `currentThread` variable, also it might be clearer with an else clause instead of a return and also clearer to avoid going through getOwner since we're already using reference directly:

``` java
public void lock() {
  Thread currentThread = Thread.currentThread();
  if (currentThread == reference.get()) {
       stamp.incrementAndGet();
  }
  else {
    while (reference.compareAndSet(null, currentThread) == false);
    stamp.set(1);
  }
}
```

One last point is whether or not to include a back-off strategy inside the while loop.

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java (hunk 11 above)
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: Is fairness a goal of this lock implementation?

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java (hunk 2 above)
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: Understood.  This is more efficient anyway. :smile:

File: src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java (hunk 11 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: I was not planning on implementing fairness as it is not free -- http://www.ibm.com/developerworks/library/j-jtp10264/. 

If we think that we should implement fairness, then we can use `AtomicReferenceArray` with an initial size. If the number of threads increases the initial size, then we can create a new instance of `AtomicReferenceArrray` and copy the contents of the old instance into the new instance.

Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: Understood.

You may want to either consider:
  o a back-off strategy in the spin loops to make sure other threads are given more opportunity to get scheduled, make progress, and let the spin loop exit (lock-free),
OR
  o using `tryLock` instead of `lock` in the implementations, and letting the `read` or `write` calls return immediately without spinning (wait-free)

Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: I get the `tryLock` option from the implementation standpoint. I think, I need to understand the backing-off strategy better.

Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/util/OptimisticReentrantLock.java]: `tryLock` option will not work for `write` calls as they have the following signatures:
- public void write(byte[] buf) throws IOException
- public void write(char[] cbuf) throws IOException

The corresponding `read` methods return an `int` indicating the number of bytes / chars read. This allows us the app to invoke the `read` calls in a loop till the necessary bytes are read. Since the `write` calls do not return any values, the app cannot spin around them.

File: README.md [hunk 12]
```diff
@@ -1,9 +1,259 @@
...
 [build-status]: https://travis-ci.org/kaazing/netx.ws
 
-WebSocket RFC-6455 implementation
+netx.ws is a Java-based implementation of [WebSocket RFC 6455](https://tools.ietf.org/html/rfc6455) client with following
+important characteristics:
+
+* Target JVM Version - Java SE 1.6
+* `java.net.Socket` and `java.net.URLConnection` style APIs
+* API for streaming binary and text payloads across WebSocket frames
+* API for sending and receiving text and binary messages
```
Comment (jfallows) [inline README.md]: This is out of scope, just temporarily right?

File: README.md [hunk 13]
```diff
@@ -1,9 +1,259 @@
...
+    <artifactId>netx.ws</artifactId>
+    <version>[0.1,1.0)</version>
+</dependency>
+```
+
+## Building this Project
+
+### Minimum requirements for building the project
+
+* Java SE Development Kit (JDK) 7 or higher
```
Comment (jfallows) [inline README.md:36]: Does it not build with JDK 6 ?
It's okay if that's really true, but might be unexpected if we are targeting Java6 APIs.  Just need to confirm please.

File: README.md [hunk 14]
```diff
@@ -1,9 +1,259 @@
...
+WsURLConnection connection = (WsURLConnection) helper.openConnection(URI.create("ws://echo.websocket.org"));
+```
+
+## Stream Binary and Text Payloads Across WebSocket Frames
+
+Both `org.kaazing.netx.ws.WebSocket` and `org.kaazing.netx.ws.WsURLConnection` support APIs for streaming both binary and text
+payloads by filtering out
+
+* specific bits related to WebSocket framing such as opcode, mask, length, etc.
+* WebSocket Control Frames such as `CLOSE`, `PING`, or `PONG` in their entirety.
```
Comment (jfallows) [inline README.md]: Recommend you describe this slightly higher level, more about what data (information) is observed than what is omitted.  Likely and application developer will think in terms of WebSocket messages, whereas you, the library developer, will think more in terms of WebSocket frames. :smile:

File: README.md [hunk 15]
```diff
@@ -1,9 +1,259 @@
...
+int charsRead = 0;
+
+while ((bytesRead != -1) && (length > 0)) {
+    charsRead = in.read(cbuf, offset, length);
+    if (charsRead != -1) {
+       offset += charsRead;
+       length -= charsRead;
+    }
+}
+```
```
Comment (jfallows) [inline README.md:129]: Worth highlighting that we handle partial UTF-8 characters properly?

File: README.md [hunk 16]
```diff
@@ -1,9 +1,259 @@
...
+Applications can send both binary and text messages using either `org.kaazing.netx.ws.WebSocket` or
+`org.kaazing.netx.ws.WsURLConnection`.
+
+### Send Binary Message Using java.io.OutputStream
+
+The following code illustrates sending a binary message using `org.kaazing.netx.ws.WebSocket` and `java.io.OutputStream`:
+
+``` java
+WebSocketFactory factory = WebSocketFactory.newInstance();
+WebSocket connection = factory.createWebSocket(URI.toString("http://echo.websocket.org"));
```
Comment (jfallows) [inline README.md]: Is `URI.toString(...)` the API call you intended here?

File: src/main/java/org/kaazing/netx/ws/WebSocket.java [hunk 17]
```diff
@@ -132,20 +132,11 @@
...
-    /**
-     * Returns a {@link MessageWriter} that can be used to send <b>binary</b> and/or <b>text</b> messages.
+     * Returns the maximum payload length that this connection will support. The default maximum payload length is 8192 bytes.
      *
-     * @return WebSocketMessageWriter   to send binary and/or text messages
-     * @throws IOException if an I/O error occurs when connecting or connection is closed
+     * @return maximum payload length for the connection
      */
-    public abstract MessageWriter getMessageWriter() throws IOException;
+    public abstract int getMaxPayloadLength();
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/WebSocket.java]: Is this intended to convey maximum WebSocket message size, or maximum WebSocket frame size?

File: src/main/java/org/kaazing/netx/ws/WebSocketFactory.java [hunk 18]
```diff
@@ -46,12 +49,14 @@
     private HttpRedirectPolicy defaultRedirectPolicy;
     private ChallengeHandler defaultChallengeHandler;
     private int defaultConnectTimeout; // milliseconds
+    private int maxPayloadLength;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/WebSocketFactory.java]: Call this `defaultMaxPayloadLength` for consistency?

File: src/main/java/org/kaazing/netx/ws/WebSocketFactory.java [hunk 19]
```diff
@@ -222,6 +236,27 @@ public void setDefaultConnectTimeout(int connectTimeout) {
...
+     * <p>
+     * If this method is invoked after a connection has been successfully established, an IllegalStateException is thrown.
+     * If the maxPayloadLength <= 0 or maxPayloadLength > Integer.MAX_VALUE - 14, an IllegalArgumentException is thrown.
+     * <p>
+     * @param maxPayloadLength  maximum payload length for the connection
+     */
+    public void setMaxPayloadLength(int maxPayloadLength) {
+        if (maxPayloadLength > Integer.MAX_VALUE - 14) {
+            throw new IllegalArgumentException(format("Maximim payload length must not exceed %d", Integer.MAX_VALUE - 14));
+        }
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/WebSocketFactory.java:252]: Hoist `Integer.MAX_VALUE - 14` as a constant, say `MAX_PAYLOAD_LENGTH_LIMIT`, for improved readability.

File: src/main/java/org/kaazing/netx/ws/internal/WebSocketInputStateMachine.java [hunk 20]
```diff
@@ -71,47 +71,53 @@ public void start(WsURLConnectionImpl connection) {
...
-        switch (state) {
-        case OPEN:
-            switch (opcode) {
-            case BINARY:
-                transition(connection, RECEIVE_BINARY_FRAME);
-                context.onBinaryReceived(frame);
+        try {
+            connection.getReadLock().lock();
+
+            DefaultWebSocketContext context = connection.getIncomingContext();
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WebSocketInputStateMachine.java:77]: Does `DefaultWebSocketContext` need to mutate as the pipeline executes?

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java [hunk 21]
```diff
@@ -178,21 +190,28 @@ public void addEnabledExtensions(String... extensions) {
...
+                    // Validate the string representation of the extension.
+                    extensionFactory.validateExtension(extension);
+                    enabledExtensions.add(extension);
+                }
+                catch (IOException ex) {
+                    // The string representation of the extension was deemed invalid by the extension.
+                    // So, it will not be negotiated during the opening handshake.
 
-                // #### TODO: Log to indicate why the enabled extension was not negotiated.
+                    // #### TODO: Log to indicate why the enabled extension was not negotiated.
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:210]: Remember to add the logging call.

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java [hunk 22]
```diff
@@ -256,70 +296,120 @@ public int getConnectTimeout() {
...
     public WsMessageReader getMessageReader() throws IOException {
-        if (messageReader == null) {
+        if (messageReader != null) {
+            return messageReader;
+        }
+
+        try {
+            stateLock.lock();
             ensureConnected();
             messageReader = new WsMessageReader(this);
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:345]: Do we need another check for non-null inside the lock here, just in case another thread racing us has already created the message reader and exited the lock.

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java [hunk 23]
```diff
@@ -329,39 +419,65 @@ public WsReader getReader() throws IOException {
...
     @Override
     public void setChallengeHandler(ChallengeHandler challengeHandler) {
+        checkConnected();
         connection.setChallengeHandler(challengeHandler);
     }
 
     @Override
     public void setConnectTimeout(int timeout) {
+        checkConnected();
         connection.setConnectTimeout(timeout);
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:478]: Is the intention that connect timeout cannot be called until after being connected?
Sorry if I misunderstood the logic.

Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:478]: Never mind, I see it now.
Recommend changing `checkConnected()` to something like `ensureReconfigurable()` which we only support in `START` state before attempting to connect.

File: src/main/java/org/kaazing/netx/ws/internal/io/IncomingSentinelExtension.java [hunk 24]
```diff
@@ -22,6 +22,7 @@
 public final class IncomingSentinelExtension extends WebSocketExtensionSpi {
     private final WebSocketFrameConsumer cachedOnBinaryReceived;
     private final WebSocketFrameConsumer cachedOnContinuationReceived;
+    private final WebSocketFrameConsumer cachedOnCloseReceived;
     private final WebSocketFrameConsumer cachedOnPingReceived;
     private final WebSocketFrameConsumer cachedOnPongReceived;
     private final WebSocketFrameConsumer cachedOnTextReceived;
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/IncomingSentinelExtension.java]: Suggest naming these as `default...` instead of `cached...`.  What do you think?

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java [hunk 25]
```diff
@@ -288,4 +288,12 @@ else if (reasonLength > 123) {
         }
     }
 
+    private void createMask(WsURLConnectionImpl connection, byte[] maskBuf) {
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: Call this `fillMask` since no creation (allocation) has occurred?

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java [hunk 26]
```diff
@@ -262,7 +262,7 @@ else if (reasonLength > 123) {
 
             int reasonOffset = closePayload.reasonOffset();
 
-            connection.getRandom().nextBytes(mask);
+            createMask(connection, mask);
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: Also, wouldn't we want to fill the mask directly into the output buffer instead of keeping it separate and copying it later?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java [hunk 27]
```diff
@@ -141,101 +145,106 @@ public int available() throws IOException {
 
     @Override
     public int read() throws IOException {
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java:146]: Are we reading 1 byte at a time and relying on the super class to fill the buffer?
That would imply we're acquiring and releasing the lock for every byte, right?

File: src/main/java/org/kaazing/netx/ws/internal/io/WsMessageReader.java [hunk 28]
```diff
@@ -188,18 +189,21 @@ public WsMessageReader(WsURLConnectionImpl connection) throws IOException {
             throw new NullPointerException(MSG_NULL_CONNECTION);
         }
 
+        int maxFrameLength = FrameUtil.calculateCapacity(false, connection.getMaxPayloadLength());
+
```
Comment (jfallows) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsMessageReader.java:191]: Consider hoisting this out to reuse recalculation for each type of reader / input stream.
Do we still need `FrameUtil` ?

File: README.md (hunk 12 above)
Comment (sanjay-saxena) [inline README.md]: That's right. I don't have a section in there describing how to use MessageReader/MessageWriter as the API and the implementation is going to change in the next/current sprint.

File: README.md (hunk 13 above)
Comment (sanjay-saxena) [inline README.md:36]: Good point. It does built with JDK 6. I will update README.md accordingly.

File: README.md (hunk 16 above)
Comment (sanjay-saxena) [inline README.md]: I meant URI.create().

File: src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java (hunk 26 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/io/OutgoingSentinelExtension.java]: We need to keep in a separate so that we can xor(^) the payload bytes appropriately.

File: src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java (hunk 27 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java:146]: Actually, we implement two read methods:
- public int read(byte[] buf, int offset, int length) throws IOException
- public int read() throws IOException

We acquire the lock in `read(byte[] buf, int offset, int length)` implementation.

Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java:146]: I will create a private method that can be called from both the `read` calls to avoid paying extra for the lock.

File: src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java [hunk 29]
```diff
@@ -258,13 +267,22 @@ else if (length == 0) {
...
+
+                return offset - mark;
+            }
+            finally {
+                stateLock.unlock();
+            }
         }
 
-        return offset - mark;
+        return 0;
```
Comment (jitsni) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java:190]: You cannot return 0 because the lock is not acquired. That's against the semantics of read() which should block until there is data or EOF. Otherwise, 0 itself will become the data.

Same goes with read(buf), read(buf, offset, length) methods. They cannot return 0 unless the buffer's length is 0. It should just block until there is data or EOF.

Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/io/WsInputStream.java:190]: Good catch!

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java [hunk 30]
```diff
@@ -207,31 +226,52 @@ public void close(int code) throws IOException {
...
             }
 
-            if (reason != null && reason.length() > 0) {
-                reasonBytes = reason.getBytes(UTF_8);
+            byte[] reasonBytes = null;
+
+            if (code != 0) {
+                // Verify code and reason against RFC 6455.
+                // If code is present, it must equal to 1000 or in range 3000 to 4999
+                if (code != 1000 && (code < 3000 || code > 4999)) {
```
Comment (jitsni) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:247]: This is not correct check as it throws exception for well defined codes (for e.g. many impls send 1002 - protocol error).  In server we do the following, when the close message is created

```
public static void validateCloseCode(int statusCode) {
    if (statusCode < 1000 || (statusCode > 1011 && statusCode <= 2999) || statusCode == 1004) {
        throw new IllegalArgumentException("Invalid close code: " + statusCode);
    }
}
```

If the close message is created from the wire, we also check for 1005 and 1006 (they are not valid on the wire.) apart from the above.

File: README.md [hunk 31]
```diff
@@ -1,9 +1,259 @@
...
+`org.kaazing.netx.ws.WebSocket` and `org.kaazing.netx.ws.WsURLConnection`.
+
+### Streaming Text Payload Using java.io.Reader
+
+The following code illustrates streaming text payloads using `org.kaazing.netx.ws.WsURLConnection` and `java.io.Reader`:
+
+``` java
+URLConnectionHelper helper = URLConnectionHelper.newInstance();
+WsURLConnection connection = (WsURLConnection) helper.openConnection(URI.create("ws://echo.websocket.org"));
+InputStream in = connection.getReader();
```
Comment (jitsni) [inline README.md]: getReader() would give Reader (not InputStream)

File: README.md [hunk 32]
```diff
@@ -1,9 +1,259 @@
...
+
+### Streaming Binary Payload Using java.io.InputStream
+
+The following code illustrates streaming binary payloads using `org.kaazing.netx.ws.WebSocket` and `java.io.InputStream`:
+
+``` java
+WebSocketFactory factory = WebSocketFactory.newInstance();
+WebSocket connection = factory.createWebSocket(URI.toString("http://echo.websocket.org"));
+InputStream in = connection.getInputStream();
+
```
Comment (jitsni) [inline README.md:90]: I think you can skip the rest of the doc once Inputstream is got. Also, the loop is wrong in this example.

File: src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java (hunk 30 above)
Comment (sanjay-saxena) [inline src/main/java/org/kaazing/netx/ws/internal/WsURLConnectionImpl.java:247]: Note that the aforementioned check is for the code coming in from the application.

**Comment by jfallows** [comment]
@sanjay-saxena, let me know if you are still planning to make further changes to `README.md`.

**Comment by sanjay-saxena** [comment]
@jfallows, no more changes coming for README.md for time being.
